src/pip/_internal/vcs/versioncontrol.py

```diff
@@ -100,6 +100,10 @@ class RemoteNotFoundError(Exception):
     pass
 
 
+class RemoteNotValidError(Exception):
+    url: str
```

Comment:
I would override `__init__()`.

```suggestion
class RemoteNotValidError(Exception):
    def __init__(self, url: str) -> None:
        super().__init__(url)
        self.url = url
```

(and change where the exception is thrown, of course)